rootfs/apply.go

```diff
@@ -0,0 +1,136 @@
+package rootfs
+
+import (
+	"io"
+	"io/ioutil"
+
+	"github.com/docker/containerd"
+	"github.com/docker/containerd/log"
+	"github.com/docker/distribution"
+	"github.com/docker/docker/pkg/archive"
+	"github.com/opencontainers/go-digest"
+	"github.com/pkg/errors"
+)
+
+type Snapshotter interface {
+	Prepare(key, parent string) ([]containerd.Mount, error)
+	Commit(name, key string) error
+	Exists(name string) bool
+}
+
+type Mounter interface {
+	Mount(mounts ...containerd.Mount) error
+	Unmount(mounts ...containerd.Mount) error
+}
+
+// ApplyLayer applies the layer to the provided parent. The resulting snapshot
+// will be stored under its ChainID.
+//
+// The parent *must* be the chainID of the parent layer.
+//
+// The returned digest is the diffID for the applied layer.
+func ApplyLayer(snapshots Snapshotter, mounter Mounter, rd io.Reader, parent digest.Digest) (digest.Digest, error) {
+	digester := digest.Canonical.Digester() // used to calculate diffID.
+	rd = io.TeeReader(rd, digester)
+
+	// create a temporary directory to work from, needs to be on same
+	// filesystem. Probably better if this shared but we'll use a tempdir, for
+	// now.
+	dir, err := ioutil.TempDir("", "unpack-")
+	if err != nil {
+		return errors.Wrapf(err, "creating temporary directory failed")
+	}
+
+	// TODO(stevvooe): Choose this key WAY more carefully. We should be able to
+	// create collisions for concurrent, conflicting unpack processes but we
+	// would need to have it be a function of the parent diffID and child
+	// layerID (since we don't know the diffID until we are done!).
+	key := dir
+
+	mounts, err := snapshots.Prepare(key, parent.String())
+	if err != nil {
+		return "", err
+	}
+
+	if err := mounter.Mount(mounts...); err != nil {
+		if err := snapshots.Rollback(key); err != nil {
+			log.L.WithError(err).Error("snapshot rollback failed")
+		}
+		return "", err
+	}
+	defer mounter.Unmount(mounts...)
+
+	if err := archive.ApplyLayer(key, rd); err != nil {
+		return "", err
+	}
+
+	diffID := digest.Digest()
+
+	chainID := diffID
+	if parent != "" {
+		chainID = ChainID([]digest.Digest{parent, chainID})
+	}
+
+	return diffID, snapshots.Commit(chainID.String(), key)
+}
+
+// Prepare the root filesystem from the set of layers. Snapshots are created
+// for each layer if they don't exist, keyed by their chain id. If the snapshot
+// already exists, it will be skipped.
+//
+// If sucessful, the chainID for the top-level layer is returned. That
+// identifier can be used to check out a snapshot.
+func Prepare(snapshots Snaphotter, mounter Mounter, layers []distribution.Descriptor,
+	// TODO(stevvooe): The following functions are candidate for internal
+	// object functions. We can use these to forumlate the beginnings of a
```

Comment:
@AkihiroSuda No. That causes too much copying internally and they end up not actually correcting for mistakes. We'll need to meticulously document these.